user-interface/src/main/java/life/qbic/datamanager/views/general/spreadsheet/Spreadsheet.java

```diff
@@ -0,0 +1,671 @@
+    if (rowNumberStyle.equals(cell.getCellStyle())) {
+      return; // does not apply to row numbers
+    }
+    if (columnHeaderStyle.equals(cell.getCellStyle())) {
+      return; // does not apply to column headers
+    }
+    Comment cellComment = createComment(errorMessage);
+    cell.setCellComment(cellComment);
+    cell.setCellStyle(invalidCellStyle);
+  }
+
+  private void markCellAsValid(Cell cell) {
+    if (!invalidCellStyle.equals(cell.getCellStyle())) {
+      return; // only apply to invalid cells
+    }
+    cell.setCellStyle(defaultCellStyle);
+    cell.removeCellComment();
+  }
+
+  private void updateCellValidationStatus(Cell cell, ValidationResult validationResult) {
+    if (validationResult.isValid()) {
+      markCellAsValid(cell);
+    } else {
+      markCellAsInvalid(cell, validationResult.errorMessage());
+    }
+  }
+
+  private boolean hasCellValidationChanged(Cell cell, ValidationResult validationResult) {
+    if (isCellValid(cell) && validationResult.isValid()) {
+      return false;
+    }
+    if (isCellInvalid(cell) && validationResult.isInvalid()) {
+      Comment existingComment = cell.getCellComment();
+      assert Objects.nonNull(validationResult.errorMessage()) : "error message is never null";
+      if (Objects.isNull(existingComment)) {
+        // error message is never null
+        return true;
+      }
+      return !validationResult.errorMessage()
+          .equals(existingComment.getString().getString());
+    }
+    return true;
+  }
+
+  private boolean isCellValid(Cell cell) {
+    return !isCellInvalid(cell);
+  }
+
+  private boolean isCellInvalid(Cell cell) {
+    return invalidCellStyle.equals(cell.getCellStyle());
+  }
+
+
+  public enum ValidationMode {
+    LAZY,
+    EAGER
+  }
+
+  private abstract class Row {
+
+  }
+
+  private final class DataRow extends Row {
+
+    private final T data;
+
+    private DataRow(T data) {
+      requireNonNull(data, "data must not be null");
+      this.data = data;
+    }
+
+    public T data() {
+      return data;
+    }
+
+    @Override
+    public boolean equals(Object object) {
+      if (this == object) {
+        return true;
+      }
+      if (object == null || getClass() != object.getClass()) {
+        return false;
+      }
+
+      DataRow dataRow = (DataRow) object;
+
+      return Objects.equals(data, dataRow.data);
+    }
+
+    @Override
+    public int hashCode() {
+      return data != null ? data.hashCode() : 0;
+    }
+
+    @Override
+    public String toString() {
+      return new StringJoiner(", ", DataRow.class.getSimpleName() + "[", "]")
+          .add("data=" + data)
+          .toString();
+    }
+  }
+
+  private class HeaderRow extends Row {
+
+  }
+
+  /**
+   * This SpreadsheetComponentFactory handles components in the spreadsheet. When a custom editor is
+   * retrieved, the editor is taken from the corresponding column.
+   */
+  private final class MyComponentFactory implements SpreadsheetComponentFactory {
```

Comment:
I would rename your implementation to something more meaningful

Reply:
Suggestion?

Reply:
How about DatamanagerSpreadsheetComponentFactory? A bit long, but maybe a subset of those words